---
title: "TCGA Biolinks download and parse data for a specific cancer"
author: "Devaki Kelkar"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: united
    highlight: tango
---

```{r setup, include=FALSE}
if (interactive() && Sys.getenv("RSTUDIO") == "") {
  source(file.path(Sys.getenv(if (.Platform$OS.type == "windows") "USERPROFILE" else "HOME"), ".vscode-R", "init.R"))
}
knitr::opts_chunk$set(echo = TRUE)
library(lintr)
# Set up custom linter

custom_linters <- list(
  assignment_linter = NULL,  # Disable assignment linter
  object_name_linter =
    lintr::object_name_linter(styles = c("snake_case", "camelCase")),
  line_length_linter = lintr::line_length_linter(120)
  # Add more custom linters as needed
)
# Run the linter on the current file
tryCatch(
  lint(filename = knitr::current_input(), linters = custom_linters),
  error = function(e) {
    message("Linting failed: ", e$message)
  }
)
```

```{r bioconductor}

if (!require("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("TCGAbiolinks")
```

```{r style document, include=FALSE}
require(styler)
styler::style_file(knitr::current_input())
```

```{r import pkgs, results='markup'}
require(TCGAbiolinks)
require(dplyr)
require(tidyr)
require(styler)
require(data.table)
require(stringr)
if (!requireNamespace("SummarizedExperiment", quietly = TRUE))
  BiocManager::install("SummarizedExperiment")
library(SummarizedExperiment)
```
## Define project of interest
```{r define project of interest}
project_name <- "TCGA-BRCA"
````
```{r query tcga-brca, results='markup'}
# Query TCGA-BRCA RNA-seq data
query <- GDCquery(
  project = project_name,
  data.category = "Transcriptome Profiling",
  data.type = "Gene Expression Quantification",
  workflow.type = "STAR - Counts"
)

# Download the data
GDCdownload(query)

# Prepare the data
data <- GDCprepare(query)

# View the first few rows of the data
head(data)

```
```{r parse data object, echo=True, results='markup', message = FALSE, warning = FALSE}
print(class(data))
# Access components of SummarizedExperiment object 'data'

# Access the assay data (gene expression counts)
# Get all available assays
assay_names <- assayNames(data)

# Create a list to store all assays
assay_data <- list()

# Loop through each assay and store it in the list
for (assay_name in assay_names) {
  assay_data[[assay_name]] <- assay(data, assay_name)
}


# Print the names of the available assays
print(paste("Available assays:", paste(assay_names, collapse = ", ")))

# Access the feature data (gene information)
feature_data <- rowData(data)
print("First few rows of feature data:")

print(head(feature_data))
sample_data <- colData(data)
print("First few rows of sample data:")
print(head(sample_data))

# Access metadata
metadata <- metadata(data)
print("Metadata:")
print(metadata)

```
# Access the sample data (clinical information)

```{r treatment data,  results='markup', message = FALSE, warning = FALSE}
# Extract treatments from sample_data
treatment <- sample_data$treatments

# Check if treatment is not NULL and has elements
if (!is.null(treatment) && length(treatment) > 0) {
  tryCatch({
    tm <- data.table::rbindlist(treatment, use.names = TRUE, fill = TRUE)
    print(head(tm))
  }, error = function(e) {
    cat("Error in rbindlist:", conditionMessage(e), "\n")
    cat("Structure of treatment:\n")
    str(treatment)
  })
} else {
  cat("Treatment data is empty or NULL\n")
  print(treatment)
}
```

```{r create rnaseq output object}
  # Try to bind the list into a data table
# Create gene ID and HGNC ID mapping
gene_id_hgnc_id <- data.frame(
  gene_id = feature_data$gene_id,
  gene_name = feature_data$gene_name,
  hgnc_id = gsub("HGNC:", "", feature_data$hgnc_id)
)

# Convert list columns to string in sample_data
sample_data[col_list] <- lapply(sample_data[col_list], 
                                function(x) sapply(x, paste, collapse = '; '))

# Check if conversion was successful
if(any(sapply(sample_data, is.list))) {
  warning("Some columns are still in list format. 
           Further investigation may be needed.")
} else {
  cat("All list columns have been successfully converted to strings.\n")
}

# Display the structure of sample_data to verify changes
col_list <- sapply(sample_data, is.list)
str(sample_data)

# Process FPKM-UQ data
fpkm_uq_data <- log2(assay_data$fpkm_uq_unstrand + 1)
fpkm_dat <- as.data.frame(t(fpkm_uq_data))
colnames(fpkm_dat) <- feature_data$gene_id
fpkm_dat$barcode <- rownames(fpkm_dat)

# Process raw counts data
raw_counts_data <- assay_data$unstranded
raw_counts_dat <- as.data.frame(t(raw_counts_data))
colnames(raw_counts_dat) <- feature_data$gene_id
raw_counts_dat$barcode <- rownames(raw_counts_dat)

# Process TPM data
tpm_data <- log2(assay_data$tpm_unstrand + 1)
tpm_dat <- as.data.frame(t(tpm_data))
colnames(tpm_dat) <- feature_data$gene_id
tpm_dat$barcode <- rownames(tpm_dat)

# Merge all data types with sample data
fpkm_dat <- merge(fpkm_dat, as.data.frame(sample_data), by = "barcode")
raw_counts_dat <- merge(raw_counts_dat, as.data.frame(sample_data),
                        by = "barcode")
tpm_dat <- merge(tpm_dat, as.data.frame(sample_data), by = "barcode")

# Display the first few rows of each data type
cat("FPKM-UQ data:\n")
cat("\nRaw counts data:\n")
print(head(fpkm_dat))
print(head(raw_counts_dat))
cat("\nTPM data:\n")
print(head(tpm_dat))
```
```{r write objects to file}
# Write gene_id_hgnc_id to CSV
write.csv(gene_id_hgnc_id, file = "gene_id_hgnc_id_mapping.csv", row.names = FALSE)

# Write FPKM data to CSV
write.csv(fpkm_dat, file = "all_fpkm_uq_data_with_clinical.csv", row.names = FALSE)

# Write raw counts data to CSV
write.csv(raw_counts_dat, file = "all_raw_counts_data_with_clinical.csv", row.names = FALSE)

# Write TPM data to CSV
write.csv(tpm_dat, file = "all_tpm_data_with_clinical.csv", row.names = FALSE)

# Write feature data to CSV
write.csv(as.data.frame(feature_data), file = "feature_data.csv", row.names = FALSE)


# Write sample data to CSV
sample_data_df <- as.data.frame(sample_data)
write.csv(sample_data_df, file = "sample_data.csv", row.names = TRUE)

print("Number of rows written to sample_data.csv:")
print(nrow(sample_data_df))


# Write metadata to CSV (if it's a data frame or can be converted to one)
if (is.data.frame(metadata) || is.list(metadata)) {
  write.csv(as.data.frame(metadata), file = "metadata.csv", row.names = FALSE)
} else {
  cat("Metadata is not in a format that can be directly written to CSV.\n")
  print(metadata)
}

cat("CSV files have been created for gene_id_hgnc_id_mapping, fpkm_data_with_clinical, 
feature_data, sample_data, and metadata (if applicable)")
```

lace```{r treatment data to csv}
# Replace "TCGA-BRCA" with your project of interest
clinical_data <- GDCquery_clinic(project = project_name, type = "clinical")
# View the column names to identify those related to treatment
colnames(clinical_data)

# Filter the columns that may contain treatment-related data
treatment_data <- clinical_data[ , grepl("treatment", colnames(clinical_data))]

# Display the treatment-related data
head(treatment_data)
write.csv(treatment_data, 'treatement_data.csv')
cat("CSV files has been created for treatment data")
```

```{r filter_rnaseq_data}
# Get the case IDs from clinical data
clinical_case_ids <- clinical_data$submitter_id

# Function to process and filter data
process_and_filter_data <- function(data, clinical_case_ids, datatype) {
  print_sample_counts <- function(data_type, df, st) {
    print(paste0(st, " samples for ", data_type, " data type: ", nrow(df)))
  }
  
  sts <- unique(data$sample_type)
  for (st in sts) {
    data <- data[data$sample_type == st, ]
    patient_repeats <- data %>% 
      group_by(patient) %>% 
      summarise(N = n()) %>% 
      ungroup()
    
    multiple_pat <- patient_repeats %>% filter(N > 1)
    sgl_pat <- patient_repeats %>% filter(N == 1)
    all_dat <- data
    
    if (nrow(multiple_pat) > 0) {
      to_average <- left_join(multiple_pat, data, by = 'patient') %>% 
        select(-N) %>% 
        group_by(patient) %>%
        summarise(across(starts_with("ENSG"), ~mean(.x, na.rm = TRUE))) %>%
        ungroup()
      sgl <- left_join(sgl_pat, data, by = 'patient')
      all_dat <- bind_rows(to_average, sgl)
    }
    
    file_name <- paste0('log2_norm_', datatype, '_clin_', gsub(' ', '_', tolower(st)), '_', str_replace_all(str_to_lower(project_name), '-', '_'), '.csv')
    write.csv(all_dat, file_name, row.names = FALSE)
    
    # Print the number of samples in each category for each data type
    print_sample_counts(data_type = datatype, df = all_dat, st = st)
  }
}

process_and_filter_data(fpkm_dat, clinical_case_ids, 'fpkm')
process_and_filter_data(raw_counts_dat, clinical_case_ids, 'raw_counts')
process_and_filter_data(tpm_dat, clinical_case_ids, 'tpm')
```






